physical/mongo.go

```diff
@@ -0,0 +1,1121 @@
+//   builtin/logical/mongodb/util.go
+//
+// Unfortunately, mgo doesn't support the ssl parameter in its MongoDB URI parsing logic, so we have to handle that
+// ourselves. See https://github.com/go-mgo/mgo/issues/84
+func (b *MongoBackend) parseMongoURI(rawUri string) (*mgo.DialInfo, error) {
+	uri, err := url.Parse(rawUri)
+	if err != nil {
+		return nil, err
+	}
+
+	info := mgo.DialInfo{
+		Addrs:    strings.Split(uri.Host, ","),
+		Database: strings.TrimPrefix(uri.Path, "/"),
+		Timeout:  10 * time.Second,
+	}
+
+	if uri.User != nil {
+		info.Username = uri.User.Username()
+		info.Password, _ = uri.User.Password()
+	}
+
+	uriSsl := false
+
+	query := uri.Query()
+	for key, values := range query {
+		var value string
+		if len(values) > 0 {
+			value = values[0]
+		}
+
+		switch key {
+		case "authSource":
+			info.Source = value
+		case "authMechanism":
+			info.Mechanism = value
+		case "gssapiServiceName":
+			info.Service = value
+		case "replicaSet":
+			info.ReplicaSetName = value
+		case "maxPoolSize":
+			poolLimit, err := strconv.Atoi(value)
+			if err != nil {
+				return nil, errors.New("bad value for maxPoolSize: " + value)
+			}
+			info.PoolLimit = poolLimit
+		case "ssl":
+			ssl, err := strconv.ParseBool(value)
+			if err != nil {
+				return nil, errors.New("bad value for ssl: " + value)
+			}
+			if ssl {
+				uriSsl = true
+			}
+		case "connect":
+			if value == "direct" {
+				info.Direct = true
+				break
+			}
+			if value == "replicaSet" {
+				break
+			}
+			fallthrough
+		default:
+			return nil, errors.New("unsupported connection URL option: " + key + "=" + value)
+		}
+	}
+
+	// deal with TLS
+	if uriSsl || b.tls {
+		tlsConfig := tls.Config{}
+
+		if b.tlsSkipVerify {
+			tlsConfig.InsecureSkipVerify = true
+		}
+
+		if b.tlsCAFile != "" {
+			caBytes, err := ioutil.ReadFile(b.tlsCAFile)
+			if err != nil {
+				return nil, errors.New("could not read CA data from '" + b.tlsCAFile + "'")
+			}
+			caPool := x509.NewCertPool()
+			ok := caPool.AppendCertsFromPEM(caBytes)
+			if !ok {
+				b.logger.Warn(fmt.Sprintf("physical/mongo: could not parse CAs from '%v'. Are you sure they are PEM encoded?", b.tlsCAFile))
+			}
+			tlsConfig.RootCAs = caPool
+		}
+
+		if b.tlsKeyFile != "" && b.tlsCertFile != "" {
+			cert, err := tls.LoadX509KeyPair(b.tlsCertFile, b.tlsKeyFile)
+			if err != nil {
+				return nil, errors.New("could not load cert and/or key from '" + b.tlsCertFile + " / '" + b.tlsKeyFile + "'")
+			}
+
+			tlsConfig.Certificates = []tls.Certificate{cert}
+		}
+
+		info.DialServer = func(addr *mgo.ServerAddr) (net.Conn, error) {
+			return tls.Dial("tcp", addr.String(), &tlsConfig)
+		}
+	}
+
+	return &info, nil
+}
+
+var keyIndex mgo.Index = mgo.Index{
+	Key:    []string{"key"},
+	Unique: true,
+}
+
+var lastCheckedInIndex mgo.Index = mgo.Index{
+	Key:         []string{"lastCheckedIn"},
+	ExpireAfter: 5 * time.Second,
+}
+
+// this part was derived from and adjusted for our needs from:
```

Comment:
It's probably not going to be possible easily. But I promise I will give it a try :)